pyuvdata/uvdata/src/miriad_wrap.pyx

```diff
@@ -0,0 +1,578 @@
+  void uvcopyvr_c (int tin, int tout)
+  void uvtrack_c  (int tno, const char *name, const char *switches)
+  void uvprobvr_c (int tno, const char *var, char *type, int *length, int *updated);
+  void uvgetvr_c  (int tno, int type, const char *var, char *data, int n);
+  void uvputvr_c  (int tno, int type, const char *var, const char *data, int n);
+  void uvselect_c (int tno, const char *object, double p1, double p2, int datasel);
+  void haccess_c(int tno, int *ihandle, const char *keyword, const char *status, int *iostat);
+  void hdaccess_c(int ihandle, int *iostat);
+
+  void hwriteb_c(int item, char buf[4], int offset, int length, int *iostat)
+  void hreadb_c(int item, char buf[4], int offset, int length, int *iostat)
+
+  void hreadi_c(int item, int *buf, int offset, int length, int *iostat)
+  void hwritei_c(int item, int *buf, int offset, int length, int *iostat)
+
+  void hwritej_c(int item, short *buf, int offset, int length, int *iostat)
+  void hreadj_c(int item, short *buf, int offset, int length, int *iostat)
+
+  void hwritel_c(int item, long *buf, int offset, int length, int *iostat)
+  void hreadl_c(int item, long *buf, int offset, int length, int *iostat)
+
+  void hwriter_c(int item, float *buf, int offset, int length, int *iostat)
+  void hreadr_c(int item, float *buf, int offset,int length, int *iostat)
+
+  void hreadd_c(int item, double *buf, int offset, int length, int *iostat)
+  void hwrited_c(int item, double *buf, int offset, int length, int *iostat)
+
+  void hwritec_c(int item, float buf[2], int offset, int length, int *iostat)
+  void hreadc_c(int item, float buf[2], int offset, int length, int *iostat)
+
+# 4 is the ITEM_HDR_SIZE in both miriad3 and 4 but cython does
+# not like using it as the length of the array size it needs to be static
+# and not defined at compile time.
+cdef extern from "io.h":
+  cdef int ITEM_HDR_SIZE "ITEM_HDR_SIZE"
+  cdef char binary_item[4]
+  cdef char real_item[4]
+  cdef char int_item[4]
+  cdef char int2_item[4]
+  cdef char int8_item[4]
+  cdef char char_item[4]
+  cdef char dble_item[4]
+  cdef char cmplx_item[4]
+
+cdef extern from "hio.h":
+  int mroundup(int a, int b)
+
+cdef extern from "maxdimc.h":
+  cpdef int _MAXCHAN "MAXCHAN"
+
+MAXCHAN = _MAXCHAN
+
+cdef numpy.int_t MAXVAR = 32768
+ctypedef numpy.int_t DTYPE_t
+ctypedef numpy.complex64_t DTYPE_c
+ctypedef numpy.float64_t DTYPE_f64
+ctypedef numpy.int64_t DTYPE_sl
+
+cdef inline int GETI(int bl):
+  return (bl - 65536) / 2048 - 1 if bl > 65536 else (bl >> 8) - 1
+
+cdef inline int GETJ(int bl):
+  return (bl - 65536) % 2048 -1 if bl > 65536 else (bl & 255) - 1
+
+cdef inline float MKBL(int i, int j):
+  return (i + 1) << 8 | (j + 1) if (i + 1 < 256 and j + 1 < 256) else (i + 1) * 2048 + (j + 1 + 65536)
+
+cdef inline void CHK_IO(int i) except *:
+  if (i != 0):
+    raise IOError("IO failed.")
+
+class MiriadError(RuntimeError):
+  pass
```

Comment:
So my read on this is that any error gets raised as a `RuntimeError`, but then nothing else happens. Is this the behavior that we want? Should we try to close the file if it's open? My guess is that this is essentially what was happening before, but we might want to think about whether this is what we'd like going forward.

Reply:
That's a pretty good point, we could wrap a lot of the function calls in try/except blocks and call a uv.close() if an error is raised.

Reply:
Also I hit plenty of errors during testing/creation that are not Runtime errors
the  `raise_miriad_error` and `except +raise_miriad_error` raise up c-level execution errors as a miriad error. So anything in the cython that will `raise` shows up as it's intended error.

turns out it is difficult to include the `e.what()` from c exceptions up into python. I tried to write a wrapper for it but couldn't figure it out and didn't want to distract myself too much.

Reply:
Thanks for looking into that. It's probably not worth investing more effort into this, especially because we weren't really handling errors properly before. Hopefully the Cython rewrite will make runtime errors even less common going forward.